_posts/2022-05-14-appsync-graphql-subscriptions.md

```diff
@@ -0,0 +1,308 @@
+
+```
+export const GET_ORDER_STATUS = gql`
+  subscription GetOrderStatus($orderId: ID) {
+    getOrderStatus(orderId: $orderId) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+```
+
+data:
+
+```
+result:{
+    data:{
+      getOrderStatus:{
+        OrderStatus: 'IN_PROGRESS',
+        orderItems: null, // orderItems is null because it was not requested in the mutation
+        createdAt: null   // createdAt is null because it was not requested in the mutation
+      }
+    }
+  }
+
+```
+
+3- Subscription arguments are indeed filters that specify when the server should notify the client.
+
+The way we have defined our subscription means that the client is interested in updates on order status when a specific orderId argument is passed through.
+
+However, if you want to enable the client to subscribe to a single order `OR` all orders, you can make this argument optional by removing the exclamation point (!) from the subscription prototype.
+
+```
+type Subscription {
+    getOrderStatus($orderId: ID): Order @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+Additionally, if you want clients to explicitly subscribe to all status updates for all orders, you should remove the argument as follows:
+
+```
+type Subscription {
+    getOrderStatus: Order
+    @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+Simply,if you want to know about all orders that are created, you could do the following:
+
+```
+type Subscription {
+    getNewOrder: Order
+        @aws_subscribe(mutations: ["addOrder"])
+}
+```
+
+The golden point here to consider is that, if the filtering field is not specified in the mutation response, the subscription is not fired! If we define the addOrder mutation as below ( createdAt filed has intentionally been removed from mutation response):
+
+```
+export const ADD_ORDER = gql`
+  mutation addOrder(orderInput: OrderInput!) {
+    addOrder(orderInput: $orderInput) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+    }
+  }
+`;
+
+```
+
+and trying to filter the subscription on createdAt filed:
+
+```
+export const ADD_ORDER_SUBSCRIPTION = gql`
+  subscription AddOrder($createdAt: string!) {
+    getNewOrder(createdAt: $createdAt) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+
+```
+
+In this case the subscription doesn't get fired because the mutation is not requesting `createdAt` which is used for filtering in the subscription data.
+
+## Conclusion
+
+We explained what AWS AppSync and GraphQL are and explored their relationship. Followed by steps on how to implement real time session between client and the server with AWS `@aws_Subscribe` directive within AppSync. And finally wrapped it up with troubleshooting, that saves you the hours of frustration and debugging!
```

Comment:
We explained
could be 
I explained

Reply:
implement real time session
should be
implement a real time session

Reply:
And finally I wrapped it up with troubleshooting, which should save you hours of frustration and debugging